controller/src/main/java/io/pravega/controller/store/stream/PersistentStreamBase.java

```diff
@@ -1420,6 +1485,87 @@ private TxnStatus handleDataNotFoundException(Throwable ex) {
+                        // we will call noteWriterMark recursively. 
+                        return Futures.exceptionallyComposeExpecting(
+                                createWriterMarkRecord(writer, timestamp, newPosition)
+                                        .thenApply(v -> WriterTimestampResponse.SUCCESS),
+                                e -> Exceptions.unwrap(e) instanceof StoreException.DataExistsException, 
+                                () -> noteWriterMark(writer, timestamp, position));
+                    } else {
+                        // sanity check and update
+                        if (record.getObject().getTimestamp() > timestamp) {
+                            // existing time is already ahead of new time
+                            return CompletableFuture.completedFuture(WriterTimestampResponse.INVALID_TIME);
+                        }
+                        if (!compareWriterPositions(record.getObject().getPosition(), newPosition)) {
+                            // existing position is already ahead of new position
+                            return CompletableFuture.completedFuture(WriterTimestampResponse.INVALID_POSITION);
+                        }
+                        
+                        // its a valid mark, update it
+                        return updateWriterMarkRecord(writer, timestamp, newPosition, record.getVersion())
+                                .thenApply(v -> WriterTimestampResponse.SUCCESS);
+                    }
+                });
+    }
+
+
+    /**
+     * Compares two given positions and returns true if position 2 is ahead of position 1. 
+     * 
+     * Note: This is not absolutely correct implementation because that would be a very costly computation. 
+     * This method simply checks if position 2 has a higher max segment number than position 1. 
+     * If those are equal, then it simply checks the offsets in each segment in position 2 to be ahead of same segment in 
+     * position 1. 
+     * It doesnt deal with complex cases where in a position 1 there could be both predecessor and successor while position 2 
+     * only contains successor. There can be many such complex checks which will require looking up ranges of segments involved
+     * which we will avoid in this method. 
+     * 
+     * @param position1 position 1 
+     * @param position2 position 2
+     * @return return true if position 2 is ahead of/or equal to position 1. false otherwise. 
+     */
+    @VisibleForTesting
+    boolean compareWriterPositions(Map<Long, Long> position1, Map<Long, Long> position2) {
+        long maxInPos2 = position2.keySet().stream().filter(position1::containsKey).max(Long::compare).orElse(Long.MIN_VALUE);
+        long maxInPos1 = position1.keySet().stream().filter(position2::containsKey).max(Long::compare).orElse(Long.MIN_VALUE);
+        
+        boolean compareMaxes = maxInPos2 >= maxInPos1;
+        
+        if (compareMaxes) {
```

Comment:
The previous implementation was better. See my above comment.

Reply:
In previous implementation, i was trying some additional checks but realized they could not be done without making this computation more costly ( as it would require fetching segment ranges from store). So i had reverted them. But there were some partial changes that had lingered on in the first iteration of the PR (like sorting the segments in the position object) which was not being used. 

Do you have suggestions for additional checks apart from that which is present now that we can include? Note: we do not want to make too many calls into store for each mark that is reported by a writer.